### ISSUE TITLE
race condition happens when using ONCE subscriptions

### ISSUE THREAD

**Issue by hellt** [karimra/gnmic#74]
Sometime ONCE response is not printed out to stdout:

The output shows two consecutive runs:
```
❯ go run . -a r1 --insecure -u admin -p admin --config /dev/null sub --path /state/system/version --mode ONCE
file_output 2020/07/07 15:42:45.454581 initialized file output: {"Cfg":{"FileName":"","FileType":"stdout"}}
gnmic 2020/07/07 15:42:45.553653 target '0.tcp.eu.ngrok.io:14852' initialized
gnmic 2020/07/07 15:42:45.555383 sending gNMI SubscribeRequest: subscribe='subscribe:{prefix:{}  subscription:{path:{elem:{name:"state"}  elem:{name:"system"}  elem:{name:"version"}}}  qos:{marking:20}  mode:ONCE}', mode='ONCE', encoding='JSON', to 0.tcp.eu.ngrok.io:14852
{
  "source": "0.tcp.eu.ngrok.io:14852",
  "subscription-name": "default",
  "timestamp": 1594129419152003783,
  "time": "2020-07-07T15:43:39.152003783+02:00",
  "prefix": "state/system/version",
  "updates": [
    {
      "Path": "version-number",
      "values": {
        "version-number": "B-20.5.R1"
      }
    },
    {
      "Path": "version-string",
      "values": {
        "version-string": "TiMOS-B-20.5.R1 both/x86_64 Nokia 7750 SR Copyright (c) 2000-2020 Nokia.\r\nAll rights reserved. All use subject to applicable license agreements.\r\nBuilt on Wed May 13 14:08:50 PDT 2020 by builder in /builds/c/205B/R1/panos/main/sros"
      }
    }
  ]
}
file_output 2020/07/07 15:42:45.753384 closing file '/dev/stdout' output



❯ go run . -a r2 --insecure -u admin -p admin --config /dev/null sub --path /state/system/version --mode ONCE
file_output 2020/07/07 15:42:49.217199 initialized file output: {"Cfg":{"FileName":"","FileType":"stdout"}}
gnmic 2020/07/07 15:42:49.352015 target '0.tcp.eu.ngrok.io:14852' initialized
gnmic 2020/07/07 15:42:49.353721 sending gNMI SubscribeRequest: subscribe='subscribe:{prefix:{}  subscription:{path:{elem:{name:"state"}  elem:{name:"system"}  elem:{name:"version"}}}  qos:{marking:20}  mode:ONCE}', mode='ONCE', encoding='JSON', to 0.tcp.eu.ngrok.io:14852
file_output 2020/07/07 15:42:49.528603 closing file '/dev/stdout' output
```

If make this a regular method call iso of goroutine, I do not see the race condition:
https://github.com/karimra/gnmic/blob/collector/collector/collector.go#L185

How crucial is it to have it a goroutine here? Since the o.Write() inside of t.Export is async anyway?

**Comment by karimra** [comment]
I saw that behavior once ( no pun intended ) but was never able to reproduce it.
It happens when the outputs are closed before the export method finishes.

The Write inside Export is async but Export itself is not async without a go routine call, the for loop would be waiting for Export to finish to get the next msg form the response channel.

2 options:

- Remove the go routine call to Export, but make the response channel buffered. Issue becomes how big of a buffer we need?
- Find a way to sync terminating c.Start() and running t.Export()

I will try to implement the second option.

**Comment by karimra** [comment]
Ended up implementing the first one but just in case of ONCE subscription

**Comment by hellt** [comment]
implemented in #71 and manually tested (2d4ff07).
